Fixes references outside function (#214)

Hoxbro opened the pull request:
Closes #214

larsoner commented:
@sebp @MatthewFlamm can you see if this fixes #214 for you?

MatthewFlamm commented:
It still fails on my usage, but I haven't compared precisely to the small test example that @Hoxbro gave in #214.  So there may be still some unresolved issue, or I'm using it incorrectly.  It does work for me on 0.8.0 however.

![image](https://user-images.githubusercontent.com/39341281/79142726-f0abbd80-7d89-11ea-8a5f-d5db6344c0da.png)

Hoxbro commented:
@MatthewFlamm Can you give me a small code example of your problem? Then I can look into it.

MatthewFlamm commented:
These days I unfortunately won't have time to try and simplify my case down to a simple example.  Here is an example from how I used it though:

https://github.com/Merck/rtdpy/blob/master/source/AxialDispersion_cc.rst
https://github.com/Merck/rtdpy/blob/master/rtdpy/ad_cc.py

Hoxbro commented:
That's okay, the linked files were more than enough. I think that I have fixed your problem to:

![image](https://user-images.githubusercontent.com/19758978/79244100-c552cc80-7e76-11ea-9e85-1a0d4a3ab9fe.png)

@larsoner The more underlying problem, seem to be that the DEDUPLICATION_TAG is not necessary placed last in the sibling_section list. I have fixed this with a for-loop but I don´t know if there is a more underlying problem here.

MatthewFlamm commented:
@Hoxbro Great work!  This works better than 0.8.0 even.  I was getting some duplication of links in 0.8.0, but it was still understandable unlike 0.9+.  Current behavior is below which works in my full code too.

![image](https://user-images.githubusercontent.com/39341281/79262731-c2001680-7e5f-11ea-9650-921cf08ab103.png)
![image](https://user-images.githubusercontent.com/39341281/79262779-cfb59c00-7e5f-11ea-91c7-d056ee885d02.png)

larsoner commented:
@Hoxbro do you see some easy way to add a test? For example, you could add something to `numpydoc/tests/tinybuild` (it's a tiny sphinx-buildable thing) that fails to do the right thing on `master`, then assert that there is some suitable substring (or regex) that shows it's working on this PR.

larsoner commented:
(the tests for `tinybuild` are in `numpydoc/tests/test_full.py`)

larsoner commented on numpydoc/tests/test_full.py:
```diff
@@ -65,3 +66,23 @@ def test_my_function(sphinx_app):
+        with open(html_file, 'r') as fid:
+            html = fid.read()
+
+        reference_list = re.findall(r'<a class="fn-backref" href="\#id\d+">(.*)<\/a>', html)
```
```suggestion
        reference_list = re.findall(r'<a class="fn-backref" href="\#id\d+">(.*)<\/a>', html)
        assert len(reference_list) == 3
```
Otherwise this could be an empty list and thus not testing what it means to test

Hoxbro replied:
I like it. It is only index.html which have 3 items in the `reference_list`. I have added the check to the test.

larsoner commented:
Awesome, thanks @Hoxbro !
